src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -76,69 +76,100 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
 
             if (featureDefinition != null)
             {
-                //
-                // Check if feature is always on
-                // If it is, result is true, goto: cache
-
-                if (featureDefinition.EnabledFor.Any(featureFilter => string.Equals(featureFilter.Name, "AlwaysOn", StringComparison.OrdinalIgnoreCase)))
+                if (featureDefinition.RequirementType == RequirementType.All && _options.IgnoreMissingFeatureFilters)
                 {
-                    enabled = true;
+                    string errorMessage = $"The 'IgnoreMissingFeatureFilters' flag cannot use used in combination with a feature of requirement type 'All'.";
+                    throw new FeatureManagementException(FeatureManagementError.MissingFeatureFilter, errorMessage);
                 }
-                else
+
+                //
+                // If the requirement type is Any, we end on a true. Requirement type All will end on a false
+                bool targetEvaluation = featureDefinition.RequirementType == RequirementType.Any;
+
+                //
+                // For all enabling filters listed in the feature's state, evaluate them according to requirement type
+                foreach (FeatureFilterConfiguration featureFilterConfiguration in featureDefinition.EnabledFor)
                 {
                     //
-                    // For all enabling filters listed in the feature's state calculate if they return true
-                    // If any executed filters return true, return true
-
-                    foreach (FeatureFilterConfiguration featureFilterConfiguration in featureDefinition.EnabledFor)
+                    // Handle AlwaysOn filters
+                    if (string.Equals(featureFilterConfiguration.Name, "AlwaysOn", StringComparison.OrdinalIgnoreCase))
                     {
-                        IFeatureFilterMetadata filter = GetFeatureFilterMetadata(featureFilterConfiguration.Name);
+                        enabled = true;
```

Comment:
`enabled = true;`

I believe having to do this comes because we start with false because of the empty feature filter case? If so, how about we special case it so we can avoid it? Something like

```
if (!featureDefinition.EnabledFor.Any())
{

}
else
{
    enabled = featureDefinition.RequirementType == RequirementType.All;
    
     foreach (...)
     {
        ...
     }
}
```

Reply:
In this particular spot, changing the logic would not affect this code much, if at all, as we still need to special case `AlwaysOn` as an expected filter that will not be found as a IFeatureFilter. `AlwaysOn` will always need to `return true` in the `Any` case and `continue` in the `All` case because continuing the code would error. 

If you check my first commit, I originally had that pattern, but it adds more `if's` and overall reads less cleanly-
1. It adds an `if` at the beginning where we default to either true or false depending on requirement type.
2. It adds an `if` for the empty list

The only benefit would only remove the `// Set enabled to true if this filter evaluated to true` section

Reply:
Not just that. You don't need the `evaluation` variable, and the implementation could really be a short circuit.

Reply:
Synced with @jimmyca15. Reverted my "default to false" changes introduced in https://github.com/microsoft/FeatureManagement-Dotnet/pull/221/commits/0ef1d89b62224b0e6c660a4c42b7c9e40c230f81 and made some readability updates.